docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+> it would not break existing usage.
+>
+> The plumbing changes would be mostly independent of the pieces that will
+> change for ABCI++ and PBTS, so working them concurrently is feasible.
+
+---
+## Detailed Design
+
+### Design Goals
+
+Specific goals of this design include:
+
+1. Remove the need for a persistent connection to each subscription client.
+   Subscribers should use the same HTTP request flow for event subscription
+   requests as for other RPC calls.
+
+2. The server retains minimal state (possibly none) per-subscriber.  In
+   particular:
+
+   - The server does not buffer unconsumed writes nor queue undelivered events
+     on a per-client basis.
+   - A client that stalls or goes idle does not cost the server any resources.
+   - Any event data that is buffered or stored is shared among _all_
+     subscribers, and is not duplicated per client.
+
+3. Slow clients have no impact (or minimal impact) on the rate of progress of
+   the consensus algorithm, beyond the ambient overhead of servicing individual
+   RPC requests.
+
+4. Clients can tell when they have missed events matching their subscription,
+   within some reasonable (configurable) window of time, and can "replay"
+   events within that window to catch up.
+
+5. Nice to have: It should be easy to use the event subscription API from
+   existing standard tools and libraries, including command-line use for
+   testing and experimentation.
+
+### Definitions
+
+- The **event stream** of a node is a single, time-ordered, heterogeneous
+  stream of event items.
+
+- Each **event item** comprises an **event datum** (for example, block header
+  metadata for a new-block event), and zero or more optional **events**.
+
+- An **event** means the [ABCI `Event` data type][abci-event], which comprises
+  a string type and zero or more string key-value **event attributes**.
+
+  The use of the new terms "event item" and "event datum" is to avert confusion
+  between the values that are published to the event bus (what we call here
+  "event items") and the ABCI `Event` data type.
+
+- The node assigns each event item a unique identifier string called a
+  **cursor**.  A cursor must be unique among all events published by a single
+  node, but it is not required to be unique globally across nodes.
+
+  Cursors are time-ordered so that given event items A and B, if A was
+  published before B, then cursor(A) < cursor(B) in lexicographic order.
+
+  A minimum viable cursor implementation is a tuple consisting of a timestamp
+  and a sequence number (e.g., `16CCC798FB5F4670-0123`). However, it may also
+  be useful to append basic type information to a cursor, to allow efficient
+  filtering (e.g., `16CCC87E91869050-0091:BeginBlock`).
+
+  The initial implementation will use the minimum viable format.
+
+### Discussion
+
+The node maintains an **event log**, a shared ordered record of the events
+published to its event bus within an operator-configurable time window.  The
```

Comment:
Are clients made aware of this configuration upon connecting to a node?

Reply:
Not as it's currently set up. Is there a use case for that? E.g., a case where the client needs to know these settings programmatically per-node?

If so, it wouldn't be too hard to add some extra bits to the node info RPC.

Reply:
I was thinking originally that the client would want to know if the node it was talking to had a window that was sufficiently long enough to include the last cursor that the client had stored but I guess in this case the client could find out by just sending the request anyway so perhaps there's not really a use case for this.

Reply:
Yeah, that's the main justification for the reply always disclosing the available range: it means the client can just ask for what it needs rather than checking first separately.